w3-filecoin.md

```diff
@@ -0,0 +1,187 @@
+# Filecoin Protocol
+
+![status:wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+## Editors
+
+- [Vasco Santos], [Protocol Labs]
+
+## Authors
+
+- [Vasco Santos], [Protocol Labs]
+
+# Abstract
+
+This spec describes the filecoin protocol used by web3.storage's "w3up" platform to interact with Spade ♠️ in order to get user/application CAR files into Filecoin deals.
+
+## Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+# Terminology
+
+## Roles
+
+There are several roles in the authorization flow:
+
+| Name        | Description |
+| ----------- | ----------- |
+| Storage Provider | [Principal] identified by [`did:web`] identifier, representing a storage provider like w3up |
+| Authority   | [Principal] that represents service provider that executes invoked capabilities |
+| Verifier   | Component of the [authority] that performs UCAN validation |
+
+### Storage Provider
+
+A _Storage Provider_ is a type of [principal] identified by a [`did:web`] identifier.
+
+A Storage Providers facilitates data storage services to applications and users, getting the requested data stored into Filecoin deals asynchronously.
+
+### Authority
+
+_Authority_ is a [principal] that executes invoked capabilities.
+
+### Verifier
+
+Component of the [authority] that performs UCAN validation
+
+# Protocol
+
+## Overview
+
+A Storage Provider is the entry point for user/application data into the web3. It will act on behalf of users to move data around into different storage points. One of the key storage points is Filecoin deals.
+
+## Authorization
+
+Storage Provider (web3.storage) must be given capabilities by a Filecoin broker Authority (Spade) to invoke capabilities to get its stored content into Filecoin deals.
+
+```mermaid
+sequenceDiagram
+    participant Authority as 🌐<br/><br/>did:web:spade.storage
+    participant Storage Provider as 🌐<br/><br/>did:web:web3.storage #32;
+
+    Authority->>Storage Provider: delegate `filecoin/add`
+```
+
+## Filecoin deals
+
+When a storage provider has enough 
+
+```mermaid
+sequenceDiagram
+    participant Authority as 🌐<br/><br/>did:web:spade.storage
+    participant Storage Provider as 🌐<br/><br/>did:web:web3.storage #32;
+
+    Storage Provider->>Authority: invoke `filecoin/add`
+    Note left of Authority: Prepare for deal
+    Authority-->>Storage Provider: receipt
+    Note left of Authority: Land deal in Filecoin
+    Authority->>Storage Provider: invoke `filecoin/deal`
+    Authority-->>Storage Provider: invoke `filecoin/fail`
```

Comment:
This part is not super clear to me. I'm assuming that spade wants to issuer receipt when request is acknowledged, yet succeed or fail it later.

If so [UCAN invocation spec has `.fx` field on the receipts](https://github.com/ucan-wg/invocation/#7-effect) to cover for this. Specifically it could point to another task in the workflow which can be then used to checked.

In other words instead of having `filecoin/deal` and `filecoin/fail` capabilities to report the status, I recommend `receipt` have a `fx.join` field pointing to e.g. `filecoin/devise` task which will eventually get receipt with either `ok` or `error`.

Reply:
Thanks, this makes total sense. I think there are two angler here:

- Confirmation receipt, which totally makes sense by your suggestion and I will add it here!
- Getting state of a deal over time. Like today we have `api.web3.storage/status/:cid` where users can check details about where given CID underlying content is stored, this is a way to give feature parity. In other words, be able to ask Spade about deal information for a given aggregate identifier.

Reply:
Also note that we can't store this information on web3.storage side given Broker will be responsible to renew deals over time and update this information

Reply:
One question @Gozala about effects. Would then `filecoin/devise` be with `issuer` and `audience` both Spade in this case?